### PR TITLE
Bounding box of tiles

### PR DESCRIPTION
**Pull request by MNimmer48**
Great package! I added a function to extract the bounding box of individual tiles which can be extracted from the subblockinfo. This kind of lower-level access was crucial to me.
I could not run tox as I do not have python 3.7 and cannot install it right now. But as I did not introduce new dependencies or the like it should hopefully be okay.

### CODE REVIEW

**Comment by heeler** [comment]
So I'm assuming the `scene_bounding_box` was insufficient? I'd be curious to better understand what you're attempting to do and how you're using the bounding boxes. I appreciate the feedback and suggestions.

File: _aicspylibczi/Reader.cpp [hunk 1]
```diff
@@ -285,6 +285,22 @@ namespace pylibczi {
       return metaSubblocks;
   }
 
+  libCZI::IntRect
+  Reader::readSubblockRect(libCZI::CDimCoordinate& plane_coord_, int index_m_)
+  {
+      SubblockSortable subBlockToFind(&plane_coord_, index_m_, isMosaic());
+      SubblockIndexVec matches = getMatches(subBlockToFind);
```
Comment (heeler) [inline _aicspylibczi/Reader.cpp:292]: You're retrieving a vector of matches. Why loop through the list and return the last one?
If the length of the list != 1 then the plane_coord_ was underspecified? Maybe throw an exception in that case or?

**Comment by MNimmer48** [comment]
Thank you for your reply, I will try to explain some more.
I have two .czi files, both have several scenes and each of those scenes consists of several hundred tiles. Both of these files are basically measurements of the same object and I need a way to match a tile from one scene to where I can find the contents of the tile in the other file. For this I have a mapping between the coordinate systems of the two files. And I need the coordinates of each individual tile, so I can apply the coordinate mapping and get the corresponding region in the second file.
`scene_bounding_box` would only give me the bounding box of the scene, not of individual tiles.

**Comment by heeler** [comment]
So If I'm following essentially Scene index and M-index should be enough to get the bounding rect for each tile. Thanks. I'll have a think and make some suggestions when I have some more time. Cheers!

**Comment by MNimmer48** [comment]
> So If I'm following essentially Scene index and M-index should be enough to get the bounding rect for each tile. Thanks. I'll have a think and make some suggestions when I have some more time. Cheers!

Yes, I think all other coordinates do not influence the YX position, so they would not need to be specified. I do not properly check for that at the moment, do you have a good way to properly check that?



> Are you needing to "draw" on the images or augment them in some way or is the reconstruction failing somehow?

Both actually. I need to reassemble tiles and the read_mosaic produces artifacts for my file. If you look closely (zoom in) at the loaded mosaic in Example 2 on the landing page, you can also see artifacts (and not just at the tile border). This may be a bug in libCZI as I don't get the artifacts if I load the tiles separately.

Thank you for your time!

File: _aicspylibczi/Reader.cpp (hunk 1 above)
Comment (MNimmer48) [reply _aicspylibczi/Reader.cpp:292]: You are right that the loop is unnecessary, as all matches should produce the same result if not misspecified. And you right that proper specification should be checked (as you wrote in your other comment).

File: aicspylibczi/CziFile.py [hunk 2]
```diff
@@ -270,6 +270,37 @@ def read_subblock_metadata(self, **kwargs):
         m_index = self._get_m_index_from_kwargs(kwargs)
 
         return self.reader.read_meta_from_subblock(plane_constraints, m_index)
+    
+    def read_subblock_rect(self, **kwargs):
+        """
+        Read the subblock specific coordinates
```
Comment (heeler) [inline aicspylibczi/CziFile.py:276]:
```suggestion
        Read the subblock specific coordinates. For non-mosaic files S only needs to be set, for mosaic files the
        S and M Dimensions need to be specified. In both cases, if underspecified only the first match is returned.
        If overspecified a PylibCZI_CDimCoordinatesOverspecifiedException is raised.
```

File: aicspylibczi/tests/test_czi_file.py [hunk 3]
```diff
@@ -237,3 +237,17 @@ def test_mosaic_image_two(data_dir, fname, expects):
...
+    ('s_3_t_1_c_3_z_5.czi', 2, -1, (39850, 39272, 475, 325)),
+    ('mosaic_test.czi', 0, 0, (0, 0, 924, 624)),
+    ('mosaic_test.czi', 0, 1, (832, 0, 924, 624)),
+])
+def test_subblock_rect(data_dir, fname, s_index, m_index, expected):
+    with open(data_dir / fname, 'rb') as fp:
+        czi = CziFile(czi_filename=fp)
+        args = {'S': s_index} if m_index < 0 else {'S': s_index, 'M': m_index}
+        data = czi.read_subblock_rect(**args)
+        assert data == expected
```
Comment (heeler) [inline aicspylibczi/tests/test_czi_file.py:253]: add a new line at the end of file. (Linting error)

File: c_tests/test_Reader.cpp [hunk 4]
```diff
@@ -320,3 +320,46 @@ TEST_CASE_METHOD(CziBgrCreator, "test_bgr_flatten", "[Reader_read_flatten_bgr]")
...
+        auto ans = answers[s_index];
+        REQUIRE( rect.x == ans.x );
+        REQUIRE( rect.y == ans.y );
+        REQUIRE( rect.w == ans.w );
+        REQUIRE( rect.h == ans.h );
+    }
+
+    libCZI::CDimCoordinate invalid_dim{{libCZI::DimensionIndex::S, 5}};
+    REQUIRE_THROWS_AS( czi->readSubblockRect(invalid_dim), pylibczi::CDimCoordinatesOverspecifiedException );
+}
```
Comment (heeler) [inline c_tests/test_Reader.cpp:365]: Add a newline here, missing EOL causes linting error.

File: aicspylibczi/tests/test_czi_file.py [hunk 5]
```diff
@@ -237,3 +237,17 @@ def test_mosaic_image_two(data_dir, fname, expects):
         rgion = (sze[0], sze[1], int(sze[2]/2), int(sze[3]/2))
         img = czi.read_mosaic(region=rgion, C=0, M=0)
     assert img.shape == expects
+
+@pytest.mark.parametrize("fname, s_index, m_index, expected", [
```
Comment (heeler) [inline aicspylibczi/tests/test_czi_file.py:241]: add another blank line here (linting again)

File: aicspylibczi/CziFile.py [hunk 6]
```diff
@@ -270,6 +270,39 @@ def read_subblock_metadata(self, **kwargs):
         m_index = self._get_m_index_from_kwargs(kwargs)
 
         return self.reader.read_meta_from_subblock(plane_constraints, m_index)
+
```
Comment (heeler) [inline aicspylibczi/CziFile.py:273]: apparently there are spaces on this line (273) which is a minor listing error. 
if you could keep the newline but remove the spaces that would be awesome.

File: aicspylibczi/CziFile.py [hunk 7]
```diff
@@ -271,6 +271,39 @@ def read_subblock_metadata(self, **kwargs):
...
+                       H = 7   # The H-dimension ("phase").
+                       V = 8   # The V-dimension ("view").
+                       M = 10  # The M_index, this is only valid for Mosaic files and must be provided here!
+
+        Returns
+        -------
+        (int, int, int, int)
+            (x, y, w, h) the bounding box of the tile
+
+        """
```
Comment (JacksonMaxfield) [inline aicspylibczi/CziFile.py:300]: @heeler have you tried rendering these docs? Do they turn out okay?

**Comment by JacksonMaxfield** [comment]
Also @heeler probably update the PR title prior to merge to be the standard `feature/support-reading-bounding-box-tiles` or similar?

**Pull request merged**